<!DOCTYPE html>
<html prefix="og: http://ogp.me/ns#">
    <head>
        <meta charset="utf-8">
        <meta http-equiv="X-UA-Compatible" content="IE=edge">
        <title>Create Krona Plot from Qiime2 Artifacts &middot; Gennaro Calendo</title>
        <meta name="description" content="Krona is a wonderful tool for interactively visualizing metagenomic samples. If you are using Qiime2 however, shaping your data into a format that can be used by Krona is not entirely straight-forward. Below, I hope to explain my workflow for creating Krona charts derived from a Qiime2 workflow. I will use the example files from the Qiime2 Moving Pictures Tutorial to show how a krona chart can be constructed from these QIIME2 artifacts.">
        <meta name="HandheldFriendly" content="True">
        <meta name="MobileOptimized" content="320">
        <meta name="generator" content="Hugo 0.58.3" />
        <meta name="robots" content="index,follow">
        <meta name="viewport" content="width=device-width, initial-scale=1.0">
        <meta property="og:title" content="Create Krona Plot from Qiime2 Artifacts">
<meta property="og:description" content="Krona is a wonderful tool for interactively visualizing metagenomic samples. If you are using Qiime2 however, shaping your data into a format that can be used by Krona is not entirely straight-forward. Below, I hope to explain my workflow for creating Krona charts derived from a Qiime2 workflow. I will use the example files from the Qiime2 Moving Pictures Tutorial to show how a krona chart can be constructed from these QIIME2 artifacts.">
<meta property="og:type" content="article">
<meta property="og:url" content="/2019/10/06/2019-10-06-create-krona-plot-from-qiime2-artifacts/">
        <link rel="stylesheet" href="/dist/styles.css">
        <link rel="stylesheet" href="https://fonts.googleapis.com/css?family=Open+Sans:300italic,400italic,600italic,700italic,400,600,700,300&subset=latin,cyrillic-ext,latin-ext,cyrillic">
        <link rel="stylesheet" href="https://maxcdn.bootstrapcdn.com/font-awesome/4.7.0/css/font-awesome.min.css" integrity="sha384-wvfXpqpZZVQGK6TAh5PVlGOfQNHSoD2xbE+QkPxCAFlNEevoEH3Sl0sibVcOQVnN" crossorigin="anonymous">
        
        
        
    </head>
    <body>
        

        <div id="wrapper">
            <header class="site-header">
                <div class="container">
                    <div class="site-title-wrapper">
                        
                            <h1 class="site-title">
                                <a title="Gennaro Calendo&#39;s Blog" href="/">Gennaro Calendo&#39;s Blog</a>
                            </h1>
                        
                        <a class="button-square" href="/index.xml"><i class="fa fa-rss"></i></a>
                        
                        
                        
                        
                            <a class="button-square button-social hint--top" data-hint="Github" title="Github" href="https://github.com/jcalendo" rel="me">
                                <i class="fa fa-github-alt"></i>
                            </a>
                        
                        
                        
                            <a class="button-square button-social hint--top" data-hint="LinkedIn" title="LinkedIn" href="https://linkedin.com/in/gennaro-calendo/" rel="me">
                                <i class="fa fa-linkedin"></i>
                            </a>
                        
                        
                        
                            <a class="button-square button-social hint--top" data-hint="Email" title="Email" href="mailto:jcalendo@gmail.com">
                                <i class="fa fa-envelope"></i>
                            </a>
                        
                    </div>

                    <ul class="site-nav">
                        
    <li class="site-nav-item">
        <a title="Blog" href="/">Blog</a>
    </li>

    <li class="site-nav-item">
        <a title="Contact" href="/page/contact/">Contact</a>
    </li>

    <li class="site-nav-item">
        <a title="About" href="/page/about/">About</a>
    </li>

                    </ul>
                </div>
            </header>

            <div id="container">


<div class="container">
    <article class="post-container" itemscope="" itemtype="http://schema.org/BlogPosting">
        <header class="post-header">
    <h1 class="post-title" itemprop="name headline">Create Krona Plot from Qiime2 Artifacts</h1>
    
    <p class="post-date">
        <span>Published <time datetime="2019-10-06" itemprop="datePublished">Sun, Oct 6, 2019</time></span>
        <span>by</span>
        <span itemscope="" itemprop="author" itemtype="https://schema.org/Person">
            <span itemprop="name">
                <a href="https://github.com/jcalendo" itemprop="url" rel="author">Gennaro Calendo</a>
            </span>
        </span>
    </p>
    
        <p class="post-reading post-line">
            <span>Estimated reading time: 5 min</span>
        </p>
    
</header>

        <div class="post-content clearfix" itemprop="articleBody">
    

    

<p><a href="https://github.com/marbl/Krona">Krona</a> is a wonderful tool for interactively visualizing metagenomic samples. If you are using Qiime2 however, shaping your data into a format that can be used by Krona is not entirely straight-forward. Below, I hope to explain my workflow for creating Krona charts derived from a Qiime2 workflow. I will use the example files from the <a href="https://docs.qiime2.org/2018.11/tutorials/moving-pictures/">Qiime2 Moving Pictures Tutorial</a> to show how a krona chart can be constructed from these QIIME2 artifacts.</p>

<h2 id="preliminary-steps">Preliminary Steps:</h2>

<h3 id="download-table-qza-taxonomy-qza-and-sample-metadata-files">Download table.qza, taxonomy.qza and sample metadata files</h3>

<ul>
<li><a href="https://data.qiime2.org/2018.11/tutorials/moving-pictures/sample_metadata.tsv">sample metadata</a></li>
<li><a href="https://docs.qiime2.org/2018.11/data/tutorials/moving-pictures/table.qza">feature table</a></li>
<li><a href="https://docs.qiime2.org/2018.11/data/tutorials/moving-pictures/taxonomy.qza">taxonomy</a></li>
</ul>

<p><em>Alternatively…</em></p>

<p>You could also download the .csv file available after creating and viewing the <code>taxa barplot</code> visualization. This csv file will contain all of the counts for every annotated ASV as well as the metadata, which is essentially what we will create using the artifacts above.</p>

<h3 id="install-the-necessary-libraries">Install the necessary libraries</h3>

<p>We will also need:</p>

<ul>
<li><code>dplyr</code> (or your preferred data manipulation package)</li>
<li><a href="https://github.com/jbisanz/qiime2R">Qiime2R</a> to easily extract data from the artifact files</li>
<li>and <a href="https://github.com/marbl/Krona/wiki/KronaTools">KronaTools</a> to create the Krona plots</li>
</ul>

<hr />

<h2 id="1-load-the-libraries">1. Load the libraries</h2>

<pre><code class="language-r">library(dplyr)
library(qiime2R)
</code></pre>

<h2 id="2-read-in-the-artifacts">2. Read in the artifacts</h2>

<pre><code class="language-r"># load metadata
md &lt;- read_tsv(&quot;sample-metadata.tsv&quot;)

# load feature table
ft &lt;- read_qza(&quot;table.qza&quot;)

# load taxonomy
tax &lt;- read_qza(&quot;taxonomy.qza&quot;)
</code></pre>

<h2 id="3-extract-data-from-the-artifacts-and-coerce-to-data-frame">3. Extract data from the artifacts and coerce to data.frame</h2>

<pre><code class="language-r"># extract data from the feature table
ft_df &lt;- ft$data %&gt;% 
  as.data.frame() %&gt;% 
  rownames_to_column(var = &quot;Feature.ID&quot;)

# extract data from taxonomy file - change factor types to characters
tax_df &lt;- tax$data %&gt;% 
  as_tibble(rownames = NULL) %&gt;% 
  mutate_if(is.factor, as.character)
</code></pre>

<h2 id="4-join-the-two-tables">4. Join the two tables</h2>

<p>The <code>Feature.ID</code> column in both tables above will function as the key that we wish to join the tables on. This will allow us to attached the taxonomy annotations to each unique Feature ID. We can also drop the Confidence column.</p>

<pre><code class="language-r">annotated_ft &lt;- ft_df %&gt;% 
  left_join(tax_df, by = &quot;Feature.ID&quot;) %&gt;% 
  select(-Confidence)
</code></pre>

<h2 id="5-gather-all-samples-to-make-a-long-form-of-the-table">5. Gather all samples to make a long form of the table</h2>

<p>Now we can begin to shape the data into the format that krona accepts. Krona expects counts in the left most column and then tab separated annotations in every other column to the right. An example of this can be found here. We can also remove some extraneous rows by filtering out counts of 0.</p>

<pre><code class="language-r">long_annotated_ft &lt;- annotated_ft %&gt;% 
  gather(key = &quot;SampleID&quot;, value = &quot;count&quot;, L1S105:L6S93) %&gt;% 
  select(SampleID, count, Taxon, -Feature.ID) %&gt;% 
  filter(count &gt; 0)
</code></pre>

<h2 id="6-join-metadata-to-the-annotated-feature-table">6. Join metadata to the annotated feature table</h2>

<p>Now that the data is in the correct form for Krona we can join the metadata to the annotated feature table in order to perform filtering functions more easily.</p>

<pre><code class="language-r">complete_df &lt;- long_annotated_ft %&gt;% 
  left_join(md, by = &quot;SampleID&quot;)
</code></pre>

<h2 id="7-separate-the-taxon-column">7. Separate the Taxon column</h2>

<p>We want each taxonomic level to be a single column in the final output tsv file therefore we have to split up the Taxon column into its separate levels. We expect a warning here because many of the annotations are not down to the species level. We also replace all NA values with an empty string so that “NA” as a string isn’t included as an annotation.</p>

<pre><code class="language-r"># define a taxonomy label vector
taxa_labels &lt;- c(&quot;Kingdom&quot;, &quot;Phylum&quot;, &quot;Class&quot;, &quot;Order&quot;, &quot;Family&quot;, &quot;Genus&quot;, &quot;Species&quot;)

# split the Taxon column into separate columns and clean up the taxon labels
split_taxon_df &lt;- complete_df %&gt;% 
  separate(Taxon, into = taxa_labels, sep = &quot;;&quot;) %&gt;% 
  mutate_at(vars(taxa_labels), str_remove, pattern = &quot;.__&quot;)  %&gt;% 
  replace(is.na(.), &quot;&quot;)
</code></pre>

<h2 id="8-filter-data-and-shape-into-krona-text-format">8. Filter data and shape into Krona text format</h2>

<p>Now that the data frame is complete we can filter for interesting groupings and then write the output to tsv. This tsv file will then be used as input for ktImportText. As an example, below I created separate dataframes based on the BodySite each sample was collected from. The goal is then to create separate Krona plots for each individual BodySite.</p>

<pre><code class="language-r">gut_df &lt;- split_taxon_df %&gt;% 
  filter(BodySite == &quot;gut&quot;) %&gt;% 
  select(count, Kingdom:Species) %&gt;% 
  write_tsv(&quot;gut-krona.tsv&quot;, col_names = FALSE)

lpalm_df &lt;- split_taxon_df %&gt;% 
  filter(BodySite == &quot;left palm&quot;) %&gt;% 
  select(count, Kingdom:Species) %&gt;% 
  write_tsv(&quot;left-palm-krona.tsv&quot;, col_names = FALSE)

rpalm_df &lt;- split_taxon_df %&gt;% 
  filter(BodySite == &quot;right palm&quot;) %&gt;% 
  select(count, Kingdom:Species) %&gt;% 
  write_tsv(&quot;right-palm-krona.tsv&quot;, col_names = FALSE)

tongue_df &lt;- split_taxon_df %&gt;% 
  filter(BodySite == &quot;tongue&quot;) %&gt;% 
  select(count, Kingdom:Species) %&gt;% 
  write_tsv(&quot;tongue-krona.tsv&quot;, col_names = FALSE)
</code></pre>

<h2 id="9-use-the-tsv-files-as-input-to-ktimporttext">9. Use the tsv files as input to ktImportText</h2>

<p>ktImportText can take multiple tsv files as input. Below we have separated the file name by the name of each plot. This will allow us to have all of the Krona plots in the same .html file.</p>

<pre><code class="language-bash">ktImportText \
  gut-krona.tsv,gut-samples \
  left-palm-krona.tsv,left-palm-samples \
  right-palm-krona.tsv,right-palm-samples \
  tongue-krona.tsv,tongue-samples \
  -o Krona-Plot.html
</code></pre>

<h2 id="view-the-resulting-krona-plot">View the resulting Krona Plot</h2>

<p><a href="/data/Krona-Plot.html">Result</a></p>

<hr />

<h2 id="the-complete-rscript">The complete rscript</h2>

<pre><code class="language-r">library(dplyr)
library(qiime2R)

# load metadata
md &lt;- read_tsv(&quot;sample-metadata.tsv&quot;)

# load feature table
ft &lt;- read_qza(&quot;table.qza&quot;)

# load taxonomy
tax &lt;- read_qza(&quot;taxonomy.qza&quot;)

# extract data from the feature table
ft_df &lt;- ft$data %&gt;% 
  as.data.frame() %&gt;% 
  rownames_to_column(var = &quot;Feature.ID&quot;)

# extract data from taxonomy file - change factor types to characters
tax_df &lt;- tax$data %&gt;% 
  as_tibble(rownames = NULL) %&gt;% 
  mutate_if(is.factor, as.character)

# Join the two tables
annotated_ft &lt;- ft_df %&gt;% 
  left_join(tax_df, by = &quot;Feature.ID&quot;) %&gt;% 
  select(-Confidence)
  
# Gather all samples to make a long form of the table
long_annotated_ft &lt;- annotated_ft %&gt;% 
  gather(key = &quot;SampleID&quot;, value = &quot;count&quot;, L1S105:L6S93) %&gt;% 
  select(SampleID, count, Taxon, -Feature.ID) %&gt;% 
  filter(count &gt; 0)

# join metadata to the annotated feature table
complete_df &lt;- long_annotated_ft %&gt;% 
  left_join(md, by = &quot;SampleID&quot;)

# define a taxonomy label vector
taxa_labels &lt;- c(&quot;Kingdom&quot;, &quot;Phylum&quot;, &quot;Class&quot;, &quot;Order&quot;, &quot;Family&quot;, &quot;Genus&quot;, &quot;Species&quot;)

# split the Taxon column into separate columns and clean up the taxon labels
split_taxon_df &lt;- complete_df %&gt;% 
  separate(Taxon, into = taxa_labels, sep = &quot;;&quot;) %&gt;% 
  mutate_at(vars(taxa_labels), str_remove, pattern = &quot;.__&quot;)  %&gt;% 
  replace(is.na(.), &quot;&quot;)

# create individual tsv files for interesting groups
gut_df &lt;- split_taxon_df %&gt;% 
  filter(BodySite == &quot;gut&quot;) %&gt;% 
  select(count, Kingdom:Species) %&gt;% 
  write_tsv(&quot;gut-krona.tsv&quot;, col_names = FALSE)

lpalm_df &lt;- split_taxon_df %&gt;% 
  filter(BodySite == &quot;left palm&quot;) %&gt;% 
  select(count, Kingdom:Species) %&gt;% 
  write_tsv(&quot;left-palm-krona.tsv&quot;, col_names = FALSE)

rpalm_df &lt;- split_taxon_df %&gt;% 
  filter(BodySite == &quot;right palm&quot;) %&gt;% 
  select(count, Kingdom:Species) %&gt;% 
  write_tsv(&quot;right-palm-krona.tsv&quot;, col_names = FALSE)

tongue_df &lt;- split_taxon_df %&gt;% 
  filter(BodySite == &quot;tongue&quot;) %&gt;% 
  select(count, Kingdom:Species) %&gt;% 
  write_tsv(&quot;tongue-krona.tsv&quot;, col_names = FALSE)
</code></pre>

</div>

        <footer class="post-footer clearfix">
    

    <div class="share">
        
            <a class="icon-twitter" href="https://twitter.com/share?text=Create%20Krona%20Plot%20from%20Qiime2%20Artifacts&url=%2f2019%2f10%2f06%2f2019-10-06-create-krona-plot-from-qiime2-artifacts%2f"
                onclick="window.open(this.href, 'twitter-share', 'width=550,height=235');return false;">
                <i class="fa fa-twitter"></i>
                <span class="hidden">Twitter</span>
            </a>
        

        
            <a class="icon-facebook" href="https://www.facebook.com/sharer/sharer.php?u=%2f2019%2f10%2f06%2f2019-10-06-create-krona-plot-from-qiime2-artifacts%2f"
                onclick="window.open(this.href, 'facebook-share','width=580,height=296');return false;">
                <i class="fa fa-facebook"></i>
                <span class="hidden">Facebook</span>
            </a>
        

        
            <a class="icon-google-plus" href="https://plus.google.com/share?url=%2f2019%2f10%2f06%2f2019-10-06-create-krona-plot-from-qiime2-artifacts%2f"
              onclick="window.open(this.href, 'google-plus-share', 'width=490,height=530');return false;">
              <i class="fa fa-google-plus"></i>
                <span class="hidden">Google+</span>
            </a>
        
        
    </div>
</footer>

        
    </article>
</div>

            </div>
        </div>

        <footer class="footer">
            <div class="container">
                <div class="site-title-wrapper">
                    <h1 class="site-title">
                        <a title="Gennaro Calendo&#39;s Blog" href="/">Gennaro Calendo&#39;s Blog</a>
                    </h1>
                    <a class="button-square button-jump-top js-jump-top" href="#">
                        <i class="fa fa-angle-up"></i>
                    </a>
                </div>

                <p class="footer-copyright">
                    <span>&copy; 2019 / Powered by <a href="https://gohugo.io/">Hugo</a></span>
                </p>
                <p class="footer-copyright">
                    <span><a href="https://github.com/roryg/ghostwriter">Ghostwriter theme</a> By <a href="http://jollygoodthemes.com">JollyGoodThemes</a></span>
                    <span>/ <a href="https://github.com/jbub/ghostwriter">Ported</a> to Hugo By <a href="https://github.com/jbub">jbub</a></span>
                </p>
            </div>
        </footer>

        <script src="/js/jquery-1.11.3.min.js"></script>
        <script src="/js/jquery.fitvids.js"></script>
        
        
            <script src="https://cdnjs.cloudflare.com/ajax/libs/highlight.js/9.12.0/highlight.min.js"></script>
        
        
        <script src="/js/scripts.js"></script>
    </body>
</html>

